VACMS-8671: Removes problematic behavioral test.

ndouglas opened the pull request:
## Description

Closes #8671.

I added a Cypress test for basic functionality and also a PHPUnit unit test to hit some of the combinations.

**EDIT**: And then factored out some logic from the block class into a separate class, and added another unit test class focusing on it.  AND I'LL DO IT AGAIN

Unfortunately, this block is a complicated class as far as mocking is concerned; it has multiple responsibilities, uses calls that conceal heavy reliance upon the Drupal service container, and the critical function we want to test is ~60 lines long.

As this is a PR that replaces a test, the impact should be minimal.  However, I did some _slight_ rewriting to support dependency injection rather than concealed `Drupal::<service>` calls, so the EWA block _does_ need some manual testing.

## QA steps
The QA steps have to wait for the frontend build to complete.  ~And right now, frontend builds aren't completing because `content-build` deleted a file 😩~ that issue is fixed, but now builds are slow/unreliable for... unknown reasons 🤔 

[Convenient link to build log](https://cms-xxmehkfeohdgqiwhxpwbeginhkgdld2s.ci.cms.va.gov/sites/default/files/build.txt)

omg it's finished building go go go

   - [x] Validate that a page with a live frontend presence has the appropriate EWA block and that the URL is linked appropriately.
   e.g. https://cms-xxmehkfeohdgqiwhxpwbeginhkgdld2s.ci.cms.va.gov/new-haven-vet-center 
![Screen Shot 2022-11-07 at 2 40 04 PM](https://user-images.githubusercontent.com/1318579/200399762-0ab1353f-a735-4d23-9224-3f2b973e3068.png)

   - [x] Validate that a page without a live frontend presence has the appropriate EWA block and that the URL is not linked at all.
   e.g. https://cms-xxmehkfeohdgqiwhxpwbeginhkgdld2s.ci.cms.va.gov/tacoma-vet-center/lacey-outstation
![Screen Shot 2022-11-07 at 2 42 34 PM](https://user-images.githubusercontent.com/1318579/200400083-f7c78f91-9102-4d35-8e40-63d8a8498db7.png)

  - [x] Validate that a page that is not supposed to have a live frontend presence has the appropriate EWA block and that there is no URL displayed. 
  e.g. https://cms-xxmehkfeohdgqiwhxpwbeginhkgdld2s.ci.cms.va.gov/central-ohio-health-care/staff-profiles/amy-lovell
![Screen Shot 2022-11-07 at 2 47 04 PM](https://user-images.githubusercontent.com/1318579/200400931-50a0eb32-6e93-4e5f-9ca6-6143a35dda77.png)

  - [x] Validate that a page with a differing number of parents (3) shows the correct behavior
  e.g. https://cms-xxmehkfeohdgqiwhxpwbeginhkgdld2s.ci.cms.va.gov/outreach-and-events/events/51074
![Screen Shot 2022-11-17 at 1 21 07 PM](https://user-images.githubusercontent.com/1318579/202526770-5178a669-0aa5-43d1-8127-def4dd5c22d4.png)

  - [x] Validate that a page with a differing number of parents (2) shows correct behavior
  e.g. https://cms-xxmehkfeohdgqiwhxpwbeginhkgdld2s.ci.cms.va.gov/node/2153/latest
![Screen Shot 2022-11-17 at 1 49 11 PM](https://user-images.githubusercontent.com/1318579/202532562-f3bfbca2-6db0-4d76-8fa1-9109ef4efc1b.png)


  - [x] Validate that a page with a differing number of parents (1) shows correct behavior
  e.g. https://cms-xxmehkfeohdgqiwhxpwbeginhkgdld2s.ci.cms.va.gov/test-data-voluptas-inventore-ut-omnis-sit-aperiam-animi-earum
![Screen Shot 2022-11-17 at 1 46 10 PM](https://user-images.githubusercontent.com/1318579/202531958-b5c1b0a9-1061-47dd-a269-b341254cc959.png)

ndouglas commented on docroot/modules/custom/va_gov_backend/modules/va_gov_backend_http_client/src/VaGovBackendHttpClient.php:
```diff
@@ -1,28 +0,0 @@
-<?php
```
This is the one, officer.

ndouglas commented on docroot/modules/custom/va_gov_backend/modules/va_gov_backend_http_client/va_gov_backend_http_client.info.yml:
```diff
@@ -1,7 +0,0 @@
-name: VA.gov Backend Http Client
```
Zapping this whole module.

ndouglas commented on docroot/modules/custom/va_gov_workflow_assignments/src/Plugin/Block/EntityMetaDisplay.php:
```diff
@@ -10,12 +10,13 @@
 use Drupal\Core\Render\RendererInterface;
 use Drupal\Core\Routing\RouteMatchInterface;
 use Drupal\Core\Url;
+use Drupal\Core\Utility\LinkGeneratorInterface;
```
Inject this rather than allow it to be retrieved via `\Drupal::service('link_generator')` call from within `\Drupal\Core\Link`.

ndouglas commented on docroot/modules/custom/va_gov_workflow_assignments/src/Plugin/Block/EntityMetaDisplay.php:
```diff
@@ -10,12 +10,13 @@
 use Drupal\va_gov_backend\Service\ExclusionTypesInterface;
 use Drupal\va_gov_backend\Service\VaGovUrlInterface;
 use Drupal\va_gov_lovell\LovellOps;
-use Drupal\va_gov_workflow_assignments\Service\EditorialWorkflowContentRepository;
+use Drupal\va_gov_workflow_assignments\Service\EditorialWorkflowContentRepositoryInterface;
```
Use interface, not implementation.

ndouglas commented on docroot/modules/custom/va_gov_workflow_assignments/src/Plugin/Block/EntityMetaDisplay.php:
```diff
@@ -127,7 +138,10 @@ public function build() {
 
     $block = $block_items = [];
-    $block_items['Content Type'] = $node->type->entity->label();
+    $block_items['Content Type'] = $this->entityTypeManager
+      ->getStorage('node_type')
+      ->load($node->bundle())
+      ->label();
```
Use calls to the entity type manager rather than magic methods on `$node`.

ndouglas commented on docroot/modules/custom/va_gov_workflow_assignments/src/Plugin/Block/EntityMetaDisplay.php:
```diff
@@ -274,10 +288,13 @@ public function getSectionHierarchyBreadcrumbLinks(NodeInterface $node) : string
    *   A section link.
    */
   private function getTermLink(TermInterface $term) {
-    return Link::fromTextAndUrl(
-      $this->t(':name', [':name' => $term->get('name')->getString()]),
+    $link = Link::fromTextAndUrl(
+      $this->t(':name', [
+        ':name' => $term->getName(),
+      ]),
       $term->toUrl()
-    )->toString();
+    );
+    return $this->linkGenerator->generateFromLink($link);
```
`Link::toString()` conceals a `Drupal::service(...)` call.

ndouglas commented on tests/behat/Drupal/ContextBase.php:
```diff
@@ -23,24 +23,6 @@ public function __construct() {
     $this->entityTypeManager = \Drupal::entityTypeManager();
   }
 
-  /**
-   * Enables the mock HTTP client.
-   *
-   * @BeforeFeature @mock_va_gov_urls
-   */
-  public static function enableVaGovBackendHttpClient() {
-    \Drupal::service('module_installer')->install(['va_gov_backend_http_client']);
-  }
-
-  /**
-   * Disables the mock HTTP client.
-   *
-   * @AfterFeature @mock_va_gov_urls
-   */
-  public static function disableVaGovBackendHttpClient() {
-    \Drupal::service('module_installer')->uninstall(['va_gov_backend_http_client']);
-  }
-
```
Installing/uninstalling this module caused a cache clear that likely caused occasional issues with concurrent tests.

ElijahLynn replied:
YAY to see this removed!

ndouglas commented on tests/behat/features/content_editing.feature:
```diff
@@ -58,52 +58,3 @@ Feature: CMS Users may effectively create & edit content
 
     # Confirm that our custom validation for Audiences & Topics is working.
     Then I should see "No more than 4 Topic/Audience tags may be selected"
-
-  @content_editing
-  Scenario: Confirm that the EWA block URL is shown correctly.
```
This test can (and should) be shifted to Cypress.  Now's a good time to do it.

ndouglas commented on tests/cypress/integration/behavioral/ewa_block.feature:
```diff
@@ -0,0 +1,21 @@
+@content_editing
+Feature: CMS Users see the EWA block.
```
A partial replacement for the removed Behat test.  This is a good target for an integration test, but we can't effectively test all of the different possibilities because of the concurrent testing apparatus.

ndouglas commented on tests/phpunit/Content/EntityMetaDisplayTest.php:
```diff
@@ -0,0 +1,238 @@
+<?php
+
```
Unit test for the EntityMetaDisplay block.

This does not offer full coverage of the possibility space.  I can expand it further, but I think this should work in concert with a refactoring of the block.  I began work on this in another branch but it grew very quickly beyond the scope of the original ticket, and that's not great.

So, chiefly, I'm trying to match the scope of the original functional tests.

ndouglas commented on docroot/modules/custom/va_gov_workflow_assignments/src/Plugin/Block/EntityMetaDisplay.php:
```diff
@@ -193,20 +200,10 @@ public function build() {
     if ($route_parameter instanceof NodeInterface) {
       return $route_parameter;
     }
-    elseif (is_numeric($route_parameter)) {
-      return $this->entityTypeManager
-        ->getStorage('node')
-        ->load($route_parameter);
-    }
-
```
This has finally been merged and shouldn't happen anymore.

ndouglas commented on docroot/modules/custom/va_gov_workflow_assignments/src/Plugin/Block/EntityMetaDisplay.php:
```diff
@@ -217,69 +214,29 @@ private function getNode() {
-
+    $links = $this->sectionHierarchyBreadcrumb->getLinksHtml($node);
+    // We exclude the first link and anything beyond the fourth.
+    $links = array_slice($links, 1, 3);
```
Split off into a separate module.

ndouglas commented on docroot/modules/custom/va_gov_workflow_assignments/src/Plugin/Block/EntityMetaDisplay.php:
```diff
@@ -217,69 +214,29 @@ private function getNode() {
     return implode(' » ', $links);
   }
 
-  /**
-   * Returns a section link.
-   *
-   * @param \Drupal\taxonomy\TermInterface $term
-   *   A taxonomy term.
-   *
-   * @return string
-   *   A section link.
-   */
-  private function getTermLink(TermInterface $term) {
-    return Link::fromTextAndUrl(
-      $this->t(':name', [':name' => $term->get('name')->getString()]),
-      $term->toUrl()
-    )->toString();
-  }
-
```
This method's calls that conceal calls to `\Drupal::service(...)`, so I've split them out, along with the above function, into a separate module that can be DI'ed more cleanly for unit testing.

swirtSJW reviewed (changes requested):
Nate this all looks amazing.  

Nice work.

I am seeing one regression  on the section breadcrumb output
https://cms-xxmehkfeohdgqiwhxpwbeginhkgdld2s.ci.cms.va.gov/outreach-and-events/events/51074
![image](https://user-images.githubusercontent.com/5752113/202513257-cfc57ee8-943c-4f3d-ad82-1abeeab7efc6.png)

ndouglas commented:
Thank you @swirtSJW!

The problem was that I switched the entity breadcrumb building part to not stop after a certain number of links, but kept the logic assuming a certain length in the block itself.

I added a new behavioral test case that shows this behavior:

![Screen Shot 2022-11-17 at 12 41 00 PM](https://user-images.githubusercontent.com/1318579/202518541-4c1923a4-f0be-4c89-bf14-961e9e2ab2de.png)

And extended the existing test case to make sure we get the expected link text.  (For some reason, I didn't add that before).

I also added some stuff to the unit tests to check the logic against varying numbers of parent terms.

I'm going to do manual testing too once this is deployed with nodes with varying numbers of parent terms just to make sure it's right.

swirtSJW reviewed (approved):
Nice adding the test in real time to catch the discovered regression.

swirtSJW commented:
Still waiting on cypress test to complete but it is a go for launch if it passes.  Really fantiastic work Nate

swirtSJW commented:
![image](https://user-images.githubusercontent.com/5752113/202544366-7c0f7bb4-49f2-42c2-a151-3f429fd0ea6f.png)
